### PR TITLE
ddl: tidb panic while query hash partition table with is null condition

### LINKED ISSUE #23802: tidb panic while query hash partition table with `is null` condition
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
CREATE TABLE `NT_HP27193` (  `COL1` int(20) DEFAULT NULL,  `COL2` varchar(20) DEFAULT NULL,  `COL4` datetime DEFAULT NULL,  `COL3` bigint(20) DEFAULT NULL,  `COL5` float DEFAULT NULL) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin PARTITION BY HASH ( `COL1`%`COL3` ) PARTITIONS 10;
select col1 from NT_HP27193 where col1 is null;
```

### 2. What did you expect to see? (Required)
query successfully 
### 3. What did you see instead (Required)
tidb panic
```sql
mysql > select col1 from NT_HP27193 where col1 is null;
(1105, 'runtime error: index out of range [1] with length 1')
```
tidb log:
```bash
[2021/04/01 15:50:02.572 +08:00] [ERROR] [conn.go:736] ["connection running loop panic"] [conn=231] [lastSQL="select col1 from NT_HP27193 where col1 is null"] [err="runtime error: index out of range [1] with length 1"] [stack="goroutine 33826544 [running]:\ngithub.com/pingcap/tidb/server.(*clientConn).Run.func1(0x4014780, 0xc032472ed0, 0xc02f879200)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:734 +0xee\npanic(0x37a7460, 0xc031e720c0)\n\t/usr/local/go/src/runtime/panic.go:679 +0x1b2\ngithub.com/pingcap/tidb/util/chunk.Row.IsNull(...)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/util/chunk/row.go:205\ngithub.com/pingcap/tidb/expression.(*Column).EvalInt(0xc028da8820, 0x406e1c0, 0xc02b02adc0, 0xc02a384a00, 0x0, 0x136136a, 0x35d81a0, 0x3938f60, 0x1)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/expression/column.go:367 +0x33b\ngithub.com/pingcap/tidb/expression.(*builtinArithmeticModIntSignedSignedSig).evalInt(0xc02a103810, 0xc02a384a00, 0x0, 0x0, 0x0, 0x0, 0x0)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/expression/builtin_arithmetic.go:1125 +0x70\ngithub.com/pingcap/tidb/expression.(*ScalarFunction).EvalInt(0xc02a3845a0, 0x406e1c0, 0xc02b02adc0, 0xc02a384a00, 0x0, 0xc0310e3e00, 0x2, 0x8, 0xc02f073920)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/expression/scalar_function.go:378 +0xf3\ngithub.com/pingcap/tidb/planner/core.(*partitionProcessor).findUsedPartitions(0x5fa3a60, 0x406e1c0, 0xc02b02adc0, 0x4057440, 0xc017dd7e40, 0x5fa4180, 0x0, 0x0, 0xc02d3bdad0, 0x1, ...)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/rule_partition_processor.go:145 +0x438\ngithub.com/pingcap/tidb/planner/core.(*partitionProcessor).pruneHashPartition(0x5fa3a60, 0x406e1c0, 0xc02b02adc0, 0x4057440, 0xc017dd7e40, 0x5fa4180, 0x0, 0x0, 0xc02d3bdad0, 0x1, ...)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/rule_partition_processor.go:207 +0x113\ngithub.com/pingcap/tidb/planner/core.(*partitionProcessor).processHashPartition(0x5fa3a60, 0xc028fad400, 0xc009373360, 0xc02a384460, 0x0, 0x409a840, 0xc02a384460)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/rule_partition_processor.go:255 +0x163\ngithub.com/pingcap/tidb/planner/core.(*partitionProcessor).prune(0x5fa3a60, 0xc028fad400, 0x10, 0x10, 0x389e040, 0x0)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/rule_partition_processor.go:555 +0x26d\ngithub.com/pingcap/tidb/planner/core.(*partitionProcessor).rewriteDataSource(0x5fa3a60, 0x4068fe0, 0xc028fad400, 0x1, 0x389e040, 0x0, 0x0)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/rule_partition_processor.go:64 +0x71d\ngithub.com/pingcap/tidb/planner/core.(*partitionProcessor).rewriteDataSource(0x5fa3a60, 0x40699e0, 0xc029e82d20, 0x0, 0xc007cdb500, 0x36cd180, 0xc007cdb500)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/rule_partition_processor.go:92 +0x62a\ngithub.com/pingcap/tidb/planner/core.(*partitionProcessor).optimize(0x5fa3a60, 0x4014780, 0xc02434f980, 0x40699e0, 0xc029e82d20, 0x40699e0, 0xc029e82d20, 0x0, 0x0)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/rule_partition_processor.go:57 +0x3f\ngithub.com/pingcap/tidb/planner/core.logicalOptimize(0x4014780, 0xc02434f980, 0x22a2, 0x40699e0, 0xc029e82d20, 0x60657b2a, 0xc029bae400, 0x13dfbc6, 0x60657b2a)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/optimizer.go:206 +0x116\ngithub.com/pingcap/tidb/planner/core.DoOptimize(0x4014780, 0xc02434f980, 0x406e1c0, 0xc02b02adc0, 0x2a2, 0x40699e0, 0xc029e82d20, 0x0, 0x0, 0x0, ...)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/optimizer.go:140 +0x77\ngithub.com/pingcap/tidb/planner.optimize(0x4014780, 0xc02434f980, 0x406e1c0, 0xc02b02adc0, 0x4033240, 0xc031853900, 0x4060060, 0xc02413f450, 0x0, 0x0, ...)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/optimize.go:289 +0x66c\ngithub.com/pingcap/tidb/planner.Optimize(0x4014780, 0xc02434f980, 0x406e1c0, 0xc02b02ad"]
```
### 4. What is your TiDB version? (Required)
```sql
tidb_version() | Release Version: v4.0.0-beta.2-2527-gd0291c802-dirty
Edition: Community
Git Commit Hash: d0291c802a8663e00530d04f6eb73a9f044dc844
Git Branch: master
UTC Build Time: 2021-03-31 19:38:58
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by Howie59**
### What problem does this PR solve?

Issue Number: close #23802 

### Check List 

Tests 

- Unit test

### Release note 

- No release note

### CODE REVIEW

**Comment by xiongjiwei** [comment]
ping! can you please resolve the confilits?

File: planner/core/integration_test.go [hunk 1]
```diff
@@ -3498,6 +3498,17 @@ func (s *testIntegrationSuite) TestIssue23736(c *C) {
...
 
+// https://github.com/pingcap/tidb/issues/23802
+func (s *testIntegrationSuite) TestPanicWhileQueryTableWithIsNull(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+
+	tk.MustExec("drop table if exists NT_HP27193")
+	tk.MustExec("CREATE TABLE `NT_HP27193` (  `COL1` int(20) DEFAULT NULL,  `COL2` varchar(20) DEFAULT NULL,  `COL4` datetime DEFAULT NULL,  `COL3` bigint(20) DEFAULT NULL,  `COL5` float DEFAULT NULL) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin PARTITION BY HASH ( `COL1`%`COL3` ) PARTITIONS 10;")
+	_, err := tk.Exec("select col1 from NT_HP27193 where col1 is null;")
+	c.Assert(err, IsNil)
```
Comment (djshow832) [inline planner/core/integration_test.go:3509]: Also insert some values and check the query result.

File: planner/core/integration_test.go [hunk 2]
```diff
@@ -3473,6 +3473,20 @@ func (s *testIntegrationSuite) TestIssue23736(c *C) {
 	c.Assert(tk.MustUseIndex("select /*+ stream_agg() */ count(1) from t0 where c > 10 and b < 2", "c"), IsFalse)
 }
 
+// https://github.com/pingcap/tidb/issues/23802
+func (s *testIntegrationSuite) TestPanicWhileQueryTableWithIsNull(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+
+	tk.MustExec("drop table if exists NT_HP27193")
+	tk.MustExec("CREATE TABLE `NT_HP27193` (  `COL1` int(20) DEFAULT NULL,  `COL2` varchar(20) DEFAULT NULL,  `COL4` datetime DEFAULT NULL,  `COL3` bigint(20) DEFAULT NULL,  `COL5` float DEFAULT NULL) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin PARTITION BY HASH ( `COL1`%`COL3` ) PARTITIONS 10;")
```
Comment (mjonss) [inline planner/core/integration_test.go:3482]: Maybe add `defer tk.MustExec("DROP TABLE NT_HP27193")`?

**Review by mjonss — APPROVED** [review]
/LGTM (but I don't yet have reviewer status).

**Review by djshow832 — APPROVED** [review]

**Review by djshow832 — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Pull request merged**